### ISSUE TITLE
Cannot create using alias and InvokableFactory

### ISSUE THREAD

**Issue by snapshotpl** [zendframework/zend-servicemanager#66]
I try to get service using alias. Alias is linked to factory with InvokableFactory. I get exception:
```
InvokableFactory requires that the requested name is provided on invocation; please update your tests or consuming container
```

After debug I have found that requested name in InvokableFactory is alias name.

It's occur in 2.7.

**Comment by weierophinney** [comment]
Which version - v2.7.2 or v3?

**Comment by snapshotpl** [comment]
@weierophinney 2.7. I have provided falling test for this case

**Comment by weierophinney** [comment]
That's *really* odd; the changes made were introduced due to testing plugin managers under identical circumstances (e.g. [zend-barcode](https://github.com/zendframework/zend-barcode/pull/16)). However, I've imported your test case locally, and can verify it. Looking into a solution now.

**Comment by weierophinney** [comment]
Oof, figured it out; it's due to the fact that v2 canonicalizes the names when storing the keys for the factories. In this particular case, the `InvokableFactory` is receiving `zendtestservicemanagertestassetinvokableobject` as the `$canonicalName`, which, if it weren't canonicalized, would be correct.

The reason the change worked in zend-barcode is due to having the `$autoAddInvokableClass` flag enabled; all the various plugins in that component are invokable classes, and, as such, when the plugin manager cannot resolve to a workable factory, that logic kicks in.

A patch will be tricky, but it's a priority for our migration, so I'll be working on it today.

**Comment by snapshotpl** [comment]
Thanks, it makes a sense

**Comment by Ocramius** [comment]
[this](https://github.com/rstgroup/zend-servicemanager/commit/ba006d69236dd5e6ef53b1c5bda557b8c89616c9#diff-4b732fbb667feb3b23722ff3ce7dd061R1237) seems wrong:

```php

    public function testCanGetServiceUsingAliasAndBuildByInvokableFactory()
    {
        $this->serviceManager->setAlias('foo', 'ZendTest\ServiceManager\TestAsset\InvokableObject');
        $this->serviceManager->setFactory('ZendTest\ServiceManager\TestAsset\InvokableObject', 'Zend\ServiceManager\Factory\InvokableFactory');
        $result = $this->serviceManager->get('foo');
        // following assertion was corrected: (you were comparing against a string)
        $this->assertInstanceOf('ZendTest\ServiceManager\TestAsset\InvokableObject', $result);
    }
```

**Comment by weierophinney** [comment]
@Ocramius Yes; found that when working on my patch locally as well. It still fails; almost have a patch ready.

**Comment by snapshotpl** [comment]
@Ocramius fail. However as @weierophinney said: still fails.

**Comment by weierophinney** [comment]
Fixed with #69.
